core-primitives/stf-executor/src/executor.rs

```diff
@@ -0,0 +1,441 @@
+		header: &PB::Header,
+		shard: &ShardIdentifier,
+		post_processing: StatePostProcessing,
+	) -> Result<ExecutedOperation>
+	where
+		PB: BlockT<Hash = H256>,
+		E: SgxExternalitiesTrait,
+	{
+		debug!("query mrenclave of self");
+		let mrenclave = self.ocall_api.get_mrenclave_of_self()?;
+		//debug!("MRENCLAVE of self is {}", mrenclave.m.to_base58());
+
+		let top_or_hash = top_or_hash::<H256>(stf_call_signed.clone(), true);
+
+		if let false = stf_call_signed.verify_signature(&mrenclave.m, &shard) {
+			error!("TrustedCallSigned: bad signature");
+			// do not panic here or users will be able to shoot workers dead by supplying a bad signature
+			return Ok(ExecutedOperation::failed(top_or_hash))
+		}
+
+		// Necessary because light client sync may not be up to date
+		// see issue #208
+		debug!("Update STF storage!");
+		let storage_hashes = Stf::get_storage_hashes_to_update(&stf_call_signed);
+		let update_map = self
+			.ocall_api
+			.get_multiple_storages_verified(storage_hashes, header)
+			.map(into_map)?;
+
+		Stf::update_storage(state, &update_map.into());
+
+		debug!("execute STF");
+		let mut extrinsic_call_backs: Vec<OpaqueCall> = Vec::new();
+		if let Err(e) = Stf::execute(state, stf_call_signed.clone(), &mut extrinsic_call_backs) {
+			error!("Stf::execute failed: {:?}", e);
+			return Ok(ExecutedOperation::failed(top_or_hash))
+		}
+
+		let call_hash = blake2_256(&stf_call_signed.encode());
+		let operation = stf_call_signed.clone().into_trusted_operation(true);
+		let operation_hash = blake2_256(&operation.encode());
+		debug!("Operation hash {:?}", operation_hash);
+		debug!("Call hash {:?}", call_hash);
+
+		if let StatePostProcessing::Prune = post_processing {
+			state.prune_state_diff();
+		}
+
+		let execution_hashes =
+			ExecutionHashes::new(H256::from(call_hash), H256::from(operation_hash));
+
+		Ok(ExecutedOperation::success(execution_hashes, top_or_hash, extrinsic_call_backs))
+	}
+}
+
+impl<OCallApi, StateHandler> StfExecuteTrustedCall for StfExecutor<OCallApi, StateHandler>
+where
+	OCallApi: EnclaveAttestationOCallApi + EnclaveOnChainOCallApi + GetStorageVerified,
+	StateHandler: HandleState,
+{
+	fn execute_trusted_call<PB>(
+		&self,
+		calls: &mut Vec<OpaqueCall>,
+		stf_call_signed: &TrustedCallSigned,
+		header: &PB::Header,
+		shard: &ShardIdentifier,
+		post_processing: StatePostProcessing,
+	) -> Result<Option<(H256, H256)>>
+	where
+		PB: BlockT<Hash = H256>,
+	{
+		// load state before executing any calls
+		let (state_lock, mut state) = self.state_handler.load_for_mutation(shard)?;
+
+		let executed_call = self.execute_trusted_call_on_stf::<PB, _>(
+			&mut state,
+			stf_call_signed,
+			header,
+			shard,
+			post_processing,
+		)?;
+
+		let (maybe_hashes, mut extrinsic_callbacks) = match executed_call.status {
+			ExecutionStatus::Success(execution_hashes, e) =>
+				(Some((execution_hashes.call_hash, execution_hashes.operation_hash)), e),
+			ExecutionStatus::Failure => (None, Vec::new()),
+		};
```

Comment:
I have kept the method signatures the same as before. The caller does not really care at the moment if a call succeeded or not. I introduced the status so that at least internally we know. And in case we decide to change the logic around it, we have that information available.